088-rebalance-progress-status.md

```diff
@@ -0,0 +1,421 @@
+
+## Motivation
+
+Information surrounding the progress of an executing partition rebalance is useful for planning future cluster operations. 
+Knowing things like how much time an ongoing partition rebalance has left to take and how much data an ongoing partition rebalance has left to transfer helps users understand the cost of an ongoing partition rebalance.
+This information helps users decide whether they should continue or cancel an ongoing rebalance, and know when future operations will be able to be safely executed.
+
+Further, having this information readily available and easily accessible via Kubernetes primitives, allows users and third-party tools like the Kubernetes CLI or [StreamsHub Console](https://github.com/streamshub/console) to easily track the progression of a partition rebalance.
+
+## Proposal
+Currently, the `KafkaRebalance` custom resource references a `ConfigMap` in the `status.optimizationResult.afterBeforeLoadConfigMap` field which stores the estimated load on the brokers before and after the optimization proposal is executed.
+This proposal extends the status section of the `KafkaRebalance` custom resource to include a `progress` section with a nested `rebalanceProgressConfigMap` field to reference that same `ConfigMap`.
+This `ConfigMap` will also be enhanced to contain information related to an ongoing partition rebalance.
+
+The `progress` section of the `KafkaRebalance` resource will look like the following:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+spec: {}
+status:
+  conditions:
+  - lastTransitionTime: "2024-11-05T15:28:23.995129903Z"
+    status: "True"
+    type: ProposalReady | Rebalancing | Stopped | Ready [1]
+  observedGeneration: 1
+  optimizationResult:
+    afterBeforeLoadConfigMap: my-rebalance
+    dataToMoveMB: 0
+    excludedBrokersForLeadership: []
+    excludedBrokersForReplicaMove: []
+    excludedTopics: []
+    intraBrokerDataToMoveMB: 0
+    monitoredPartitionsPercentage: 100
+    numIntraBrokerReplicaMovements: 0
+    numLeaderMovements: 16
+    numReplicaMovements: 0
+    onDemandBalancednessScoreAfter: 95.4347095948149
+    onDemandBalancednessScoreBefore: 89.4347095948149
+    provisionRecommendation: ""
+    provisionStatus: RIGHT_SIZED
+    recentWindows: 1
+  progress: [1]
+    rebalanceProgressConfigMap: my-rebalance [2]
+```
+[1] The `progress` section will be visible during the `ProposalReady`, `Rebalancing`, `Stopped`, `NotReady` and `Ready` states.
+
+[2] The `ConfigMap` containing information related to the ongoing partition rebalance.
+
+In the `ConfigMap`, we will add the following fields:
+- **estimatedTimeToCompletionInMinutes**: The estimated amount time it will take in minutes until partition rebalance is complete.
+- **completedDataMovementPercentage**: The percentage of the data movement of the partition rebalance that is completed e.g. values in the range [0-100].
```

Comment:
What is data movement? Is that percentage of bytes? Or percentage of partition-replicass being moved? I.e., if I have proposal moving 20 partition replicas with total amount of 1Gi of data and right now we have moved 16 partitions with 200Mi of data ... will the percentage be 80% or 20%? I think the name should make this clear.

Also, will it be an integer? Or will it be more precise? If it is integer, how will it be rounded?

Reply:
The data movement here is the percentage of bytes, maybe we could rename to `completedByteMovementPercentage` to make this more clear, would that be suitable?

> Also, will it be an integer? Or will it be more precise? If it is integer, how will it be rounded?

It will be an integer and rounded down to be more conservative. I'll update the proposal with this information.

Reply:
`completedByteMovementPercentage` sounds good as well as integrer rounded down. Please make sure it is included int he proposal.

Reply:
Updated